tests/phpunit/url_rewriter_html_head_setup_test.php

```diff
@@ -0,0 +1,209 @@
+ * @copyright  2017 Catalyst IT Australia {@link http://www.catalyst-au.net}
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+use local_cleanurls\local\cleaner\cleaner;
+use local_cleanurls\local\uncleaner\uncleaner;
+use local_cleanurls\url_rewriter;
+
+defined('MOODLE_INTERNAL') || die();
+require_once(__DIR__ . '/cleanurls_testcase.php');
+
+/**
+ * @package    local_cleanurls
+ * @author     Daniel Thee Roperto <daniel.roperto@catalyst-au.net>
+ * @copyright  2017 Catalyst IT Australia {@link http://www.catalyst-au.net}
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class local_cleanurls_url_rewriter_html_head_setup_test extends local_cleanurls_testcase {
+    public function tearDown() {
+        parent::tearDown();
+        url_rewriter::clear_last_apache_note();
+    }
+
+    /**
+     * Mocks 4 different types of access to this page.
+     *
+     * Pages with 'nonrouted' are pages that are not coming through a router (using unclean URL).
+     * Pages with 'routed' are coming through a router (using clean URL).
+     *
+     * The 'routed-canonical' is a page with the correct clean url.
+     * The 'routed-legacy' is a page with a clean url that works but should be replaced by its canonical version.
+     *
+     * @return array
+     */
+    public function provider_with_page_mocks() {
+        return [
+            'nonrouted-uncleanable' => [(object)['routed' => false, 'cleanable' => false]],
+            'nonrouted-cleanable'   => [(object)['routed' => false, 'cleanable' => true]],
+            'routed-canonical'      => [(object)['routed' => true, 'cleanable' => false]],
+            'routed-legacy'         => [(object)['routed' => true, 'cleanable' => true]],
+        ];
+    }
+
+    /**
+     * Sets up the page using the given data provider information.
+     *
+     * @param $mock
+     */
+    protected function mock_page($mock) {
+        global $CFG, $ME, $ORIGINALME, $PAGE;
+
+        $PAGE = new stdClass();
+
+        if ($mock->routed) {
+            $ME = '/local/cleanurls/tests/foo.php';
+            $ORIGINALME = $mock->cleanable ? '/local/cleanurls/tests/oldbar' : '/local/cleanurls/tests/bar';
+            $CFG->uncleanedurl = (new moodle_url($ME))->raw_out();
+        } else {
+            unset($CFG->uncleanedurl);
+            unset($ORIGINALME);
+            $ME = $mock->cleanable ? '/local/cleanurls/tests/foo.php' : '/local/cleanurls/tests/legacy.php';
+        }
+
+        $PAGE->url = new moodle_url($ME);
+    }
+
+    /**
+     * This only ensures we have the basic test URLs used in other tests.
+     *
+     * @dataProvider provider_with_page_mocks
+     */
+    public function test_it_requires_some_test_urls_for_testing($mock) {
+        $unclean = null;
+
+        $expectedclean = 'http://www.example.com/moodle/local/cleanurls/tests/bar';
+
+        if (!$mock->routed && !$mock->cleanable) {
```

Comment:
if logic inside tests is a code smell to me. Why not move these urls into the data provider?